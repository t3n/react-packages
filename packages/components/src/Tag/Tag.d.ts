import React from 'react';
import { MarginProps } from 'styled-system';
export declare type TagColorVariant = 'primary' | 'secondary' | 'inverse' | 'black' | 'highlight' | 'warning';
export interface TagProps extends MarginProps {
    colorVariant?: TagColorVariant;
    link?: string;
    small?: boolean;
    icon?: JSX.Element;
    onClick?: () => void;
}
export declare const Tag: React.FC<TagProps>;
