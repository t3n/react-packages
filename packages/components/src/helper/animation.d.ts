import { ThemeProps } from '@t3n/theme';
export declare const backgroundAnimation: import("styled-components").FlattenInterpolation<import("styled-components").ThemedStyledProps<ThemeProps, any>>;
