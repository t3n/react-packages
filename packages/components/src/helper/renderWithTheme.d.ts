import 'jest-styled-components';
import React from 'react';
import { RenderOptions } from '@testing-library/react';
import { Theme } from '@t3n/theme';
interface OptionalTheme {
    theme?: Theme;
}
export declare const renderWithTheme: (ui: React.ReactElement, { theme, ...options }: RenderOptions & OptionalTheme) => import("@testing-library/react").RenderResult<typeof import("@testing-library/dom/types/queries"), HTMLElement>;
export {};
