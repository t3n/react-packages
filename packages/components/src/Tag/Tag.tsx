import React from 'react';
import styled, { css } from 'styled-components';
import { color, margin, MarginProps, padding, variant } from 'styled-system';

import { composeTextStyle, Theme } from '@t3n/theme';

export type TagColorVariant =
  | 'primary'
  | 'secondary'
  | 'inverse'
  | 'black'
  | 'highlight'
  | 'warning'
  | 'notice'
  | 'success'
  | 'error';

export interface TagProps extends MarginProps {
  colorVariant?: TagColorVariant;
  link?: string;
  small?: boolean;
  icon?: JSX.Element;
  onClick?: () => void;
}

interface StyledTagProps {
  theme: Theme;
  variant: TagColorVariant;
  link?: string;
  small: boolean;
  clickable: boolean;
}

const StyledTag = styled.div.attrs((props: StyledTagProps) => ({
  href: props.link ? props.link : undefined,
}))`
  display: inline-flex;
  align-items: center;
  border-radius: 30px;
  ${margin};
  ${({ theme, small }: StyledTagProps) =>
    composeTextStyle({ theme, textStyle: small ? 'small' : 'regular' })}
  ${({ theme, small }: StyledTagProps) =>
    padding({ theme, px: small ? 2 : 3, py: small ? '2px' : 1 })}

  ${variant({
    variants: {
      primary: {
        bg: 'background.primary',
        color: 'text.primary',
      },
      secondary: {
        bg: 'background.secondary',
        color: 'text.primary',
      },
      inverse: {
        bg: 'background.inverse',
        color: 'text.inverse',
      },
      black: {
        bg: 'shades.black',
        color: 'text.inverse',
      },
      highlight: {
        bg: 'background.highlight',
        color: 'text.inverse',
      },
      error: {
        bg: 'background.highlight',
        color: 'text.inverse',
      },
      warning: {
        bg: 'feedback.warn',
        color: 'text.primary',
      },
      notice: {
        bg: 'feedback.notice',
        color: 'text.primary',
      },
      success: {
        bg: 'feedback.success',
        color: 'text.primary',
      },
    },
  })}

  svg {
    ${({ theme }) => margin({ theme, mt: 1, ml: 2 })}
    fill: ${({ theme, variant: variantStyle }: StyledTagProps) =>
      variantStyle === 'inverse' ||
      variantStyle === 'black' ||
      variantStyle === 'highlight' ||
      variantStyle === 'error'
        ? theme.colors.text.inverse
        : theme.colors.text.primary}
  }

  ${({ theme, clickable }: StyledTagProps) =>
    clickable &&
    css`
      text-decoration: none;
      &:hover {
        cursor: pointer;
        ${color({ theme, color: 'brand.white', bg: 'brand.red' })};

        svg {
          fill: ${theme.colors.brand.white};
        }
      }
    `}
`;

const Tag: React.FC<TagProps> = ({
  children,
  link,
  colorVariant,
  onClick,
  icon,
  small,
  ...rest
}) => (
  <StyledTag
    as={link ? 'a' : 'div'}
    link={link}
    clickable={!!link || !!onClick}
    variant={colorVariant || 'secondary'}
    onClick={onClick}
    small={small || false}
    {...rest}
  >
    {children}
    {icon}
  </StyledTag>
);

export default Tag;
