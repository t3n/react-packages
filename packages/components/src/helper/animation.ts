import { css, keyframes } from 'styled-components';

import { ThemeProps } from '@t3n/theme';

const backgroundAnimationKeyframes = ({ theme }: ThemeProps) => keyframes`
  0% {
    background-color: ${theme.colors.shades.grey204}
  }

  50% {
    background-color: ${theme.colors.shades.grey244}
  }

  100% {
    background-color: ${theme.colors.shades.grey204}
  }
`;

// eslint-disable-next-line import/prefer-default-export
export const backgroundAnimation = css`
  animation: ${backgroundAnimationKeyframes} 2s linear infinite;
`;
