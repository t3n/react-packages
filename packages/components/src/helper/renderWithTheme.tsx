/* eslint-disable import/no-extraneous-dependencies */
import 'jest-styled-components';

import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { Theme, theme as originalTheme } from '@t3n/theme';

interface OptionalTheme {
  theme?: Theme;
}

// eslint-disable-next-line import/prefer-default-export
export const renderWithTheme = (
  ui: React.ReactElement,
  { theme, ...options }: RenderOptions & OptionalTheme
) => {
  // merge over default theme

  const mergedTheme = { ...originalTheme, ...theme };

  const Wrapper: React.FC = ({ children }) => (
    <ThemeProvider theme={mergedTheme}>{children}</ThemeProvider>
  );

  return render(ui, { wrapper: Wrapper, ...options });
};
