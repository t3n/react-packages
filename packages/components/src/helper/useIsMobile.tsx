import { Theme } from '@t3n/theme';
import { useCallback, useContext, useEffect, useState } from 'react';
import { ThemeContext } from 'styled-components';

const useIsMobile = (): boolean => {
  const [isMobile, setIsMobile] = useState(false);

  const theme: Theme = useContext(ThemeContext);
  const fontSize = window
    .getComputedStyle(window.document.body)
    .getPropertyValue('font-size');
  const mobileBreakpoint =
    parseInt(theme.breakpoints[1], 10) * parseInt(fontSize, 10);

  const testIsMobile = useCallback(() => {
    if (window.matchMedia(`(max-width: ${mobileBreakpoint}px)`).matches) {
      setIsMobile(true);
    } else {
      setIsMobile(false);
    }
  }, [mobileBreakpoint]);

  useEffect(() => {
    testIsMobile();
    window.addEventListener('resize', testIsMobile);

    return () => {
      window.removeEventListener('resize', testIsMobile);
    };
  }, [testIsMobile]);

  return isMobile;
};

export default useIsMobile;
