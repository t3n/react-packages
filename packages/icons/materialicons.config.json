{
  "renameRules": {
    "3dRotation": "ThreeDRotation"
  },
  "ignoreFolders": [],
  "defaultStyle": "filled"
}
