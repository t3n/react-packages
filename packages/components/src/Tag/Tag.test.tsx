import React from 'react';

import { MaterialClear } from '@t3n/icons';

import { renderWithTheme } from '../helper/renderWithTheme';
import { Tag } from './Tag';

test('Tag Variants matches snapshot', () => {
  const { container } = renderWithTheme(
    <>
      <Tag colorVariant="primary">primary tag</Tag>
      <Tag colorVariant="secondary">secondary tag</Tag>
      <Tag colorVariant="inverse">inverse tag</Tag>
      <Tag colorVariant="black">black tag</Tag>
      <Tag colorVariant="primary" small>
        small primary tag
      </Tag>
      <Tag colorVariant="secondary" small>
        small secondary tag
      </Tag>
      <Tag colorVariant="inverse" small>
        small inverse tag
      </Tag>
      <Tag colorVariant="black" small>
        small black tag
      </Tag>
    </>,
    {}
  );

  expect(container).toMatchSnapshot();
});

test('Icon is rendered within the tag', () => {
  const { container } = renderWithTheme(
    <Tag colorVariant="primary" icon={<MaterialClear />}>
      Tag with Icon
    </Tag>,
    {}
  );

  expect(container.getElementsByTagName('svg')).not.toBeNull();
});
