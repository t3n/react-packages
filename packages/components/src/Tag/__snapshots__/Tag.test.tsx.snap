// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag Variants matches snapshot 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 30px;
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 1rem;
  line-height: 1.5em;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
  background-color: #fff;
  color: #2a2a2a;
}

.c0 svg {
  margin-top: 4px;
  margin-left: 8px;
  fill: #2a2a2a;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 30px;
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 1rem;
  line-height: 1.5em;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
  background-color: #f4f4f4;
  color: #2a2a2a;
}

.c1 svg {
  margin-top: 4px;
  margin-left: 8px;
  fill: #2a2a2a;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 30px;
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 1rem;
  line-height: 1.5em;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
  background-color: #2a2a2a;
  color: #fff;
}

.c2 svg {
  margin-top: 4px;
  margin-left: 8px;
  fill: #fff;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 30px;
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 1rem;
  line-height: 1.5em;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
  background-color: #000;
  color: #fff;
}

.c3 svg {
  margin-top: 4px;
  margin-left: 8px;
  fill: #fff;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 30px;
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5em;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 2px;
  padding-bottom: 2px;
  background-color: #fff;
  color: #2a2a2a;
}

.c4 svg {
  margin-top: 4px;
  margin-left: 8px;
  fill: #2a2a2a;
}

.c5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 30px;
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5em;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 2px;
  padding-bottom: 2px;
  background-color: #f4f4f4;
  color: #2a2a2a;
}

.c5 svg {
  margin-top: 4px;
  margin-left: 8px;
  fill: #2a2a2a;
}

.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 30px;
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5em;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 2px;
  padding-bottom: 2px;
  background-color: #2a2a2a;
  color: #fff;
}

.c6 svg {
  margin-top: 4px;
  margin-left: 8px;
  fill: #fff;
}

.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 30px;
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5em;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 2px;
  padding-bottom: 2px;
  background-color: #000;
  color: #fff;
}

.c7 svg {
  margin-top: 4px;
  margin-left: 8px;
  fill: #fff;
}

@media screen and (min-width:30rem) {
  .c0 {
    line-height: 1.5em;
  }
}

@media screen and (min-width:48rem) {
  .c0 {
    line-height: 1.666em;
  }
}

@media screen and (min-width:30rem) {
  .c1 {
    line-height: 1.5em;
  }
}

@media screen and (min-width:48rem) {
  .c1 {
    line-height: 1.666em;
  }
}

@media screen and (min-width:30rem) {
  .c2 {
    line-height: 1.5em;
  }
}

@media screen and (min-width:48rem) {
  .c2 {
    line-height: 1.666em;
  }
}

@media screen and (min-width:30rem) {
  .c3 {
    line-height: 1.5em;
  }
}

@media screen and (min-width:48rem) {
  .c3 {
    line-height: 1.666em;
  }
}

<div>
  <div
    class="c0"
  >
    primary tag
  </div>
  <div
    class="c1"
  >
    secondary tag
  </div>
  <div
    class="c2"
  >
    inverse tag
  </div>
  <div
    class="c3"
  >
    black tag
  </div>
  <div
    class="c4"
  >
    small primary tag
  </div>
  <div
    class="c5"
  >
    small secondary tag
  </div>
  <div
    class="c6"
  >
    small inverse tag
  </div>
  <div
    class="c7"
  >
    small black tag
  </div>
</div>
`;
